### PR TITLE
Handle early wakeup of counter loop

### LINKED ISSUE #56695: Strange events from System.Runtime EventCounters under Linux
Hi.
After migrating from .net 5 sdk 5.0.101 to net 5 sdk 5.0.302, I noticed strange behavior in System.Runtime EventCounters under Linux. It's about the "cpu-usage" counter. I began to receive two events in the required time interval. Moreover, in the second event, the value is clearly incorrect.

I wrote a small program to demonstrate this behavior:
```csharp
using System;
using System.Collections.Generic;
using System.Diagnostics.Tracing;

namespace EventCountersProblem
{
    class Program
    {
        static void Main(string[] args)
        {
            var eventListener = new CpuTimeEventListener();
            
            Console.WriteLine("Press enter for exit");
            Console.ReadLine();
            GC.KeepAlive(eventListener);
        }
    }

    public class CpuTimeEventListener : EventListener
    {
        private readonly EventLevel _level = EventLevel.Verbose;
        
        protected override void OnEventSourceCreated(EventSource source)
        {
            if (source.Name.Equals("System.Runtime"))
            {
                var refreshInterval = new Dictionary<string, string>();
                refreshInterval.Add("EventCounterIntervalSec", "5");
                EnableEvents(source, _level, (EventKeywords)(-1), refreshInterval);
            }
        }
        protected override void OnEventWritten(EventWrittenEventArgs eventData)
        {
            if (eventData.EventName.Equals("EventCounters"))
            {
                for (int i = 0; i < eventData.Payload.Count; i++)
                {
                    IDictionary<string, object> eventPayload = eventData.Payload[i] as IDictionary<string, object>;

                    if (eventPayload != null)
                    {
                        if (eventPayload["Name"] == "cpu-usage")
                        {
                            Console.WriteLine($"{DateTime.Now}:");
                            foreach (var ep in eventPayload)
                            {
                                Console.WriteLine($"{ep.Key}: {ep.Value}");
                            }

                            Console.WriteLine("=========");
                        }
                    }
                }
            }
        }
    }
}

```
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

</Project>

```
This project is published as a self-contained application:
```sh
dotnet publish -f net5.0 -r linux-x64 -c Release
```

Below is the output of the program when working under Windows:
```
Press enter for exit
01.08.2021 15:56:44:
Name: cpu-usage
DisplayName: CPU Usage
Mean: 0
StandardDeviation: 0
Count: 1
Min: 0
Max: 0
IntervalSec: 5,0044165
Series: Interval=5000
CounterType: Mean
Metadata:
DisplayUnits: %
=========
01.08.2021 15:56:49:
Name: cpu-usage
DisplayName: CPU Usage
Mean: 0
StandardDeviation: 0
Count: 1
Min: 0
Max: 0
IntervalSec: 5,00996
Series: Interval=5000
CounterType: Mean
Metadata:
DisplayUnits: %
=========
```
Everything is correct here, one event every 5 seconds with the correct value.

Below is the output of the program when running under Linux Centos (CentOS Linux release 7.4.1708 (Core)):
(When run in wsl2 (ubuntu or centos), the behavior is the same)
```
8/1/2021 3:11:48 PM:
Name: cpu-usage
DisplayName: CPU Usage
Mean: 0
StandardDeviation: 0
Count: 1
Min: 0
Max: 0
IntervalSec: 4.9997354
Series: Interval=5000
CounterType: Mean
Metadata:
DisplayUnits: %
=========
8/1/2021 3:11:48 PM:
Name: cpu-usage
DisplayName: CPU Usage
Mean: 24
StandardDeviation: 0
Count: 1
Min: 24
Max: 24
IntervalSec: 0.0314531
Series: Interval=5000
CounterType: Mean
Metadata:
DisplayUnits: %
=========
8/1/2021 3:11:53 PM:
Name: cpu-usage
DisplayName: CPU Usage
Mean: 0
StandardDeviation: 0
Count: 1
Min: 0
Max: 0
IntervalSec: 4.968951
Series: Interval=5000
CounterType: Mean
Metadata:
DisplayUnits: %
=========
8/1/2021 3:11:53 PM:
Name: cpu-usage
DisplayName: CPU Usage
Mean: 35
StandardDeviation: 0
Count: 1
Min: 35
Max: 35
IntervalSec: 0.0021859
Series: Interval=5000
CounterType: Mean
Metadata:
DisplayUnits: %
=========

```

Here we see that two events are generated in a 5 second interval. In the first event, the correct value. In the second event, the value is incorrect (24% and 35% of cpu-usage).
The second event has a similarly strange meaning in IntervalSec field.

Please tell me what could be the reason for this behavior? Is this a bug? Is this related to #53836?

Thanks in advance for your reply.

p.s.
My dotnet --info output below:
```
Пакет SDK для .NET (отражающий любой global.json):
 Version:   6.0.100-preview.6.21355.2
 Commit:    7f8e0d76c0

Среда выполнения:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-preview.6.21355.2\

Host (useful for support):
  Version: 6.0.0-preview.6.21352.12
  Commit:  770d630b28

.NET SDKs installed:
  3.1.410 [C:\Program Files\dotnet\sdk]
  5.0.301 [C:\Program Files\dotnet\sdk]
  5.0.302 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.6.21355.2 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.6.21355.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.6.21352.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-preview.6.21353.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

```

### PR DESCRIPTION
**Pull request by josalem**
Fix #56695
* prevent a negative delta value

Copy of comment from #56695:

> I think I've worked out what happened.
> 
> The new logic:
> 
> ```cs
> TimeSpan delta = now - _nextPollingTimeStamp;
> if (delta > TimeSpan.Zero && _pollingIntervalInMilliseconds > 0)
> 	_nextPollingTimeStamp += TimeSpan.FromMilliseconds(_pollingIntervalInMilliseconds * Math.Ceiling(delta.TotalMilliseconds / _pollingIntervalInMilliseconds));
> ```
> 
> can have wonky behavior if the sleep in `CounterGroup.PollForValues` wakes up early.
> 
> `delta` can be negative, which will result in `_nextPollingTimeStamp` not being updated. This means that `sleepDurationInMilliseconds` will be set to `1` and the counter will fire again on the next loop because `_nextPollingTimeStamp < now + new TimeSpan.FromMilliseconds(1)` will be true.
> 
> https://github.com/dotnet/runtime/blob/4d26112c07391f563cb0ca46af1993635399007e/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs#L267-L306
> 
> A fix would be to switch to the following:
> 
> ```cs
> private const TimeSpan MinimumPollingInterval = TimeSpan.FromSeconds(1);
> 
> // ...
> 
> TimeSpan delta = now - _nextPollingTimeStamp;
> delta = MinimumPollingInterval > delta ? MinimumPollingInterval : delta;
> if (_pollingIntervalInMilliseconds > 0)
> 	_nextPollingTimeStamp += TimeSpan.FromMilliseconds(_pollingIntervalInMilliseconds * Math.Ceiling(delta.TotalMilliseconds / _pollingIntervalInMilliseconds));
> ```
> 
> The `now - _nextPollingTimeStamp` is calculating the difference between when the counter woke up (`now`) and when it was _supposed_ to wake up (`_nextPollingTimeStamp`). This is supposed to then increment `_nextPollingTimeStamp` to the next discrete increment of `_pollingIntervalInMillseconds` from the origin time. This prevents the interval from sliding if counters take sufficiently long amounts of time to calculate. This updated logic without the loop, however, only increments if `now > _nextPollingTimeStamp` (if the counter woke up _after_ the intended timestamp).

We'll need to backport these changes to 3.1 and 5.0 if they are approved.

### CODE REVIEW

**Review by mikem8361 — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs [hunk 1]
```diff
@@ -252,7 +253,8 @@ private void OnTimer()
                 {
                     _timeStampSinceCollectionStarted = now;
                     TimeSpan delta = now - _nextPollingTimeStamp;
-                    if (delta > TimeSpan.Zero && _pollingIntervalInMilliseconds > 0)
+                    delta = s_minimumPollingInterval > delta ? s_minimumPollingInterval : delta;
```
Comment (hoyosjs) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:256]: delta = _pollingIntervalInMilliseconds > delta ? _pollingIntervalInMilliseconds  : delta

**Review by hoyosjs — CHANGES_REQUESTED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs [hunk 2]
```diff
@@ -23,6 +23,7 @@ internal sealed class CounterGroup
         private readonly EventSource _eventSource;
         private readonly List<DiagnosticCounter> _counters;
         private static readonly object s_counterGroupLock = new object();
+        private static readonly TimeSpan s_minimumPollingInterval = TimeSpan.FromSeconds(1);
```
Comment (hoyosjs) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:26]:
```suggestion
```

**Review by hoyosjs — APPROVED** [review]

**Comment by josalem** [comment]
CI doesn't seem to be starting for some reason...

**Review by davmason — APPROVED** [review]

**Comment by hoyosjs** [comment]
Failuer is not unique to this PR (see #57174)

**Comment by hoyosjs** [comment]
Created 
https://github.com/dotnet/runtime/issues/57375

**Pull request merged**